plugins/WindowManager/TopLevelWindowModel.cpp

```diff
@@ -721,8 +724,29 @@ void TopLevelWindowModel::activateTopMostWindowWithoutId(int forbiddenId)
     }
 }
 
-void TopLevelWindowModel::activateNullWindow()
+void TopLevelWindowModel::rootFocus(bool focus)
+{
+    if (focus) {
+        // Give focus back to previous focused window, only if null window is focused
+        // If null window is not focused, a different app had taken the focus and we
+        // should repect that or if a pendigActivation is going on
+        if (m_previousWindow && !m_previousWindow->focused() && !m_pendingActivation &&
+            m_nullWindow == m_focusedWindow && m_previousWindow != m_nullWindow) {
+            m_previousWindow->activate();
+        }
+    } else {
+        if (!m_nullWindow->focused()) {
+            m_nullWindow->activate();
+        }
+    }
+}
+
+// Pending Activation will block refocus of previous focused window
+// this is needed since a surface activation with miral is async
+// and a activation of placeholder is sync. This causes a race condition
+// between placeholder and prev window. This results in prev window
+// gets focused as it will always be later then placeholder.
```

Comment:
Some grammar fixes needed in this comment block.

`since a surface activation is async` -> `since surface activation is async,`
`and a activation of placeholder` -> `and activation of the placeholder`
`gets focused as` -> `getting focused, as`
`later then` -> `later than`